### PR TITLE
Numbers: positional names and alignments

### LINKED ISSUE #14: Support named holes starting with digit
Confirmed to be working in Serilog (Thanks @nblumhardt ), so we need it also :)

So this is OK:

```
{0a}
{0000a}
```

### PR DESCRIPTION
**Pull request by jods4**
fixes #14

### CODE REVIEW

**Comment by 304NotModified** [comment]
ps: http://cc.davelozinski.com/c-sharp/fastest-way-to-convert-a-string-to-an-int

``` c#
for (int x = 0; x < NumberOfIterations; x++)
{
        y = 0;
        for (int i = 0; i < s[x].Length; i++)
               y = y * 10 + (s[x][i] - '0');
        total += y;
}
```

or `Convert.ToInt32`

**Comment by jods4** [comment]
Yeah I can rewrite using this for maximum perf, skipping the string copy entirely on the way.
Need to be a little more robust wrt to error handling, though. The sample code doesn't check `s[x] < '9'`.
Do we need to support spaces around the number?

**Comment by 304NotModified** [comment]
> Do we need to support spaces around the number?

Yes, see #4

**Comment by jods4** [comment]
lol and tests fail because I have an uncomitted file, I'm such a noob.

**Comment by 304NotModified** [comment]
LOL just change git history and nobody would see it ;)

**Comment by jods4** [comment]
#4 does seem to test `{x, -4 }`. My own test seem to indicate this is accepted.

**Comment by 304NotModified** [comment]
yes space after number looks supported

File: src/Parser/TemplateParser.cs [hunk 1]
```diff
@@ -94,7 +95,7 @@ private void ParseCloseBracketPart()
         private void ParseHole(HoleType type)
         {
             string name = ReadUntil(HoleDelimiters);
-            string aligment = Peek() == ',' ? ParseAlignment() : null;
+            int? aligment = Peek() == ',' ? ParseAlignment() : (int?)null;
```
Comment (304NotModified) [inline src/Parser/TemplateParser.cs:97]: if this is an `if else`, you don't need the cast. But it this also more performant? (as casts are "slow")

File: src/Parser/TemplateParser.cs [hunk 2]
```diff
@@ -94,7 +95,7 @@ private void ParseCloseBracketPart()
...
             string format = Peek() == ':' ? ParseFormat() : null;
```
Comment (304NotModified) [inline src/Parser/TemplateParser.cs:98]: We need an else case here (performance)

File: src/Parser/TemplateParser.cs [hunk 3]
```diff
@@ -106,11 +107,15 @@ private string ParseFormat()
...
-        private string ParseAlignment()
+
+        private int ParseAlignment()
         {
             Skip(',');
-            //tood don't parse non-numeric?
-            return ReadUntil(AlignmentDelimiters);
+            string text = ReadUntil(AlignmentDelimiters);
+            int alignment;
+            if (!int.TryParse(text, NumberStyles.Integer, CultureInfo.InvariantCulture, out alignment))
```
Comment (304NotModified) [inline src/Parser/TemplateParser.cs]: isn't ordinalCulture better in this case? (performance and syntax isn't culture aware)

Comment (jods4) [inline src/Parser/TemplateParser.cs]: There is no `OrdinalCulture` you confuse with string comparison options. (Code is removed anyway).

File: src/Parser/TemplateParser.cs (hunk 1 above)
Comment (jods4) [inline src/Parser/TemplateParser.cs:97]: There is no cast at runtime, this is just to make the C# compiler happy.

File: src/Parser/TemplateParser.cs [hunk 4]
```diff
@@ -124,10 +123,47 @@ private int ParseAlignment()
...
 
+        private void SkipSpaces()
+        {
+            // Can be out of bounds, but never in correct use (inside a hole).
+            while (_template[_position] == ' ') _position++;
+        }
+
+        private int ReadInt()
+        {               
+            SkipSpaces();
```
Comment (304NotModified) [inline src/Parser/TemplateParser.cs:161]: We have to test for tabs ;)

**Comment by jods4** [comment]
I think the alignment parsing is ready to be merged.

File: src/Parser/TemplateParser.cs (hunk 4 above)
Comment (jods4) [inline src/Parser/TemplateParser.cs:161]: In my testing, `string.Format("{0,\t2}", 1)` throws...
If we want to support that, the question becomes: what is a space? The whole Unicode stuff?
I think that for syntax inside the alignment, `' '` is vastly sufficient.

Comment (304NotModified) [inline src/Parser/TemplateParser.cs:161]: thanks will add it to the cases.

PS: {0a} should work (non-numeric)

**Comment by jods4** [comment]
@304NotModified all ready to merge I think.

**Comment by 304NotModified** [comment]
Looks great! Thanks again!

**Pull request merged**